Feature: add recompletion schedule start date option

rhell4 opened the pull request:
This feature allows a recompletion schedule to start on a specific date. For example you set up the schedule to recomplete every 3 months but want this to start 1st Jan, you can set up the schedule any time and set the start date to 1st Jan instead of having to set it up on that day.

rhell4 committed:
feature: add recompletion schedule start date option

danmarsden commented:
What happens when someone puts a date from the past into that field? - eg 1st April 2024 instead of 2025 - the next cron run would trigger recompletion which may be unexpected.

I think we need to have some handling to prevent the "nextruntime" to be set to a date in the past or make it really clear in the UI that they are setting a date that might cause strange behaviour (eg run recompletion now) - will it then set the next date using the period plus the original run time? - so we end up resetting today, then resetting again next cron run for multiple periods until we get to the current date?

sorry about the conflicts too - In the process of reviewing this one I picked up some of the other PR's sitting here for a while.

thanks!

danmarsden commented:
on top of that - lets watch out for a course restore where this value is pulled through into a new course - what happens then?

Peterburnett commented:
@rhell4 Wherever we land with historical stuff, lets wrap it in a testcase to ensure that its well understood and predictable

rhell4 commented:
Hey @danmarsden,

Good catch, I didn't think about users selecting a previous date. I've made the following changes to make things more clear and prevent any unexpected behaviour;
1. Updated the help string to include 'This date must be a current or future date.'
2. Updated the form validation so it only accepts current or future dates
    - This will provide the error message 'Invalid schedule start date entered, this must be a future or current date.' if an earlier date is selected
3. Added a `startyear` to the field so that the user can't accidentally select a previous year
    - Unfortunately this is the best we can do with the core `date_selector` element without creating our own custom element, but I thought it was better than nothing 
4. Added unit tests to test the form validation

For the backup/restore concerns, this setting should have no impact. This new setting only alters the initial `nextresettime` config that is saved on the form submission and isn't saved as it's own config. This means this setting isn't included in the backup/restore process.

Hopefully that all makes sense, let me know what you think or if you have any other concerns.

Thanks!

Peterburnett commented:
I've just given this a once over briefly. I think locking this to the future is the easiest path forwards here as discussed, and avoids a whole class of complexity for a relatively small use case

danmarsden commented:
Thanks - looks good to me - nice to have the unit tests too - I've cherry-picked into the 403_STABLE branch too.
